File: original-python-solutions/breadth-first-search.py
from collections import deque

graph = {
  "you": ["alice", "bob", "claire"],
  "alice": ["anuj", "peggy"],
  "bob": ["peggy"],
  "claire": ["thom", "jonny"],
  "anuj": [],
  "peggy": [],
  "thom": [],
  "jonny": []
}

def breadth_first_search(name):
  search_deque = deque()
  search_deque += graph[name]
  searched = []

  while search_deque:
    person = search_deque.popleft()

    if not person in searched:
      if person_is_seller(person):
        print(person + " is a mango seller!")

        return True
      else:
        search_deque += graph[person]
        
        searched.append(person)

  return False

def person_is_seller(name):
  return name == "jonny"

breadth_first_search("you")

File: algorithms/breadth-first-search.js
import { Queue, Graph } from '../data-sctructures/index.js';

export function breadthFirstSearch(graph, source, compareFn) {
  const queue = new Queue();
  const checked = [];

  queue.enqueue(...graph.getNeighbors(source));

  while (!queue.isEmpty()) {
    const value = queue.dequeue();

    if (checked.includes(value)) continue;

    if (compareFn(value)) {
      return value;
    }

    checked.push(value);

    queue.enqueue(...graph.getNeighbors(value));
  }
}

const graph = new Graph(
  ['you', 'alice', 'claire', 'anuj', 'peggy', 'josh'],
  [
    { source: 'you', destination: 'alice' },
    { source: 'you', destination: 'claire' },
    { source: 'alice', destination: 'anuj' },
    { source: 'alice', destination: 'peggy' },
    { source: 'claire', destination: 'peggy' },
    { source: 'claire', destination: 'josh' }
  ]
);

console.log(
  breadthFirstSearch(graph, 'you', node => node.endsWith('sh')),
  'is a mango seller!'
);
